File: threader/src/main/java/com/walhalla/boilerplate/threading/MainThreadImpl.java
package com.walhalla.boilerplate.threading;

import android.os.Handler;
import android.os.Looper;

import com.walhalla.boilerplate.domain.executor.MainThread;

public class MainThreadImpl implements MainThread {

    private static MainThread sThread;

    private Handler mHandler;

    public MainThreadImpl() {
        mHandler = new Handler(Looper.getMainLooper());
    }

    @Override
    public void post(Runnable runnable) {
        mHandler.post(runnable);
    }

    public static MainThread getInstance() {
        if (sThread == null) {
            sThread = new MainThreadImpl();
        }

        return sThread;
    }
}

File: threader/src/main/java/com/walhalla/boilerplate/threading/MainThreadImpl.kt
package com.walhalla.boilerplate.threading

import android.os.Handler
import android.os.Looper
import com.walhalla.boilerplate.domain.executor.MainThread

class MainThreadImpl : MainThread {

    private val mHandler: Handler = Handler(Looper.getMainLooper())

    override fun post(runnable: Runnable) {
        mHandler.post(runnable)
    }

    companion object {
        private var sThread: MainThread? = null

        @JvmStatic
        val instance: MainThread
            get() {
                if (sThread == null) {
                    sThread = MainThreadImpl()
                }
                return sThread!!
            }
    }
}
